enhancements/etcd/automated-backups.md

```diff
@@ -0,0 +1,376 @@
+
+### Saving across all nodes
+
+One method is to spread out the saved backups on all available control-plane nodes. This would involve:
+- Scanning all control-plane nodes for existing backup files
+- Pruning the backup file from the node with the oldest backup
+- Saving a new backup on the node that has the oldest most-recent-backup
+
+E.g given the following nodes N(x) with backups B(t) where x is the node suffix and t is the backup timestamp:
+- N1[B1, B5], N2[B2, B4], N3[B3, B6]
+- We would select N1 to prune since B1 is the oldest of (B1,B2,B3)
+- We would select N2 as the node to save the latest backup on since B4 is the oldest of (B4,B5,B6).
+
+This approach has the benefit of spreading out backups across nodes which reduces the likelihood of losing the most recent backup in the event of losing access to a node. However it also has the following drawbacks:
+
+- Gathering the state of the backup files is more complicated:
+  - The backup controller could serially schedule a pod on each node to try and gather the backup state. It's unclear how the controller would gather that state (e.g have each pod write to a configmap).
+- Serially scheduling on each node to gather the state increases the time to perform a backup
+- If a node is unhealthy and is unschedulable then the backup may be stalled as we need to gather the state by scheduling on all nodes
+- Providing visibility into the state of the backup files for the admin is non-trivial:
+  - We would require a status object that lists all nodes and the ordered list of backup filenames per node.
+
+### Saving to a local PV
+
+Alternatively the backup files can be saved at a single location on cluster on a [local Persistent Volume](https://kubernetes.io/docs/concepts/storage/volumes/#local). Saving to a single volume for local storage has the following pros as compared to distributing backups across the control-plane nodes:
+
+- Simplifies the retention logic for locally stored backups. This reduces the maintenance burden of having different retention strategies for locally saved backups vs off-cluster (e.g nfs or gcePersistentDisk volumes).
+- Using a PV also grants the flexibility to do lifecycle management of backups e.g what to do with the old backups if a new schedule is created for a different volume.
+- The saved backups state is more visible from a single location by looking at the PV contents vs looking at all nodes to piece the order of saved backups.
+- For local backups if the cluster is unhealthy the automated backups won't be stalled:
+  - If another node besides the one with the local volume is unavailable we can still save the backup.
+  - If that node with the local volume is inaccessible, we can still provision a new volume on another available node and update the spec to save backups to the new location. 
+- Can leverage [volume snapshots](https://kubernetes.io/docs/concepts/storage/volume-snapshots/) or volume cloning to improve availability of saved backups.
```

Comment:
Added a section debating the merits of whether we want to distribute and prune backups across all nodes, vs save them to a local volume. Depending on what we decide to pick I'll update the API accordingly (e.g specifying a PVC in the spec).
I'm leaning towards the local PV approach but @williamcaban @deads2k @tjungblu please share your thoughts or concerns.

Reply:
Seems like thomas made some comments on the commit but github won't show them. Recapping from my notifications here:

```
@tjungblu: that's only available for CSI drivers:

> VolumeSnapshot support is only available for CSI drivers.

I'm not sure if the local storage is based on CSI nowadays? @jsafrane sorry to pull you into that discussion - do you know if that's available?

@jsafrane: LocalVolumes are not based on CSI and do not support snapshot.
Only on SNO clusters we support [LVMO](https://docs.openshift.com/container-platform/4.12/storage/persistent_storage/persistent_storage_local/persistent-storage-using-lvms.html) and LVM snapshots.
```

Seems like volume snapshots won't be possible for local volumes at least. Not sure if there is another way to replicate/clone the volume. Or write to a secondary local volume as an extra backup. If not, then that is a considerable risk to lose the local volume if the node becomes inaccessible.

Reply:
My main concern is that the local PV object requires a dedicated disk or partition (both of which are not part of the standard OCP requirements) and the local PV also requires for the cluster to be operational to access the data. How can these constraints be overcome?

Reply:
if local volumes work ootb, I don't mind using them over host mounts.  I think you'd still have to save across nodes.  I don't see local volumes a substitute for general PVs if that's a later requirement.